### PR TITLE
Updated URLScan View

### PR DESCRIPTION
**Pull request by RajuKoushik**
*Passing all the tests

Signed-off-by: Raju Koushik g.rajukoushik@gmail.com

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -3,6 +3,7 @@ psycopg2==2.7.1
 djangorestframework==3.6.3
 celery==4.0.2
 redis==2.10.5
+giturl.py
```
Comment (pombredanne) [inline requirements.txt]: Do you need a specific version?

File: scanapp/tasks.py [hunk 2]
```diff
@@ -24,21 +24,26 @@
 from __future__ import absolute_import, unicode_literals
 
 import json
```
Comment (pombredanne) [inline scanapp/tasks.py:26]: Sort your imports

File: scanapp/tasks.py [hunk 3]
```diff
@@ -57,6 +62,42 @@ def scan_code_async(url, scan_id, path, file_name):
 
 
 @app.task
+def scan_code_async_final(url, scan_id):
+    """
+    Create and save a file at `path` present at `URL` using `scan_id` and bare `path`
```
Comment (pombredanne) [inline scanapp/tasks.py]: Where is `path`? this is not an arg?

File: scanapp/tasks.py [hunk 4]
```diff
@@ -57,6 +62,42 @@ def scan_code_async(url, scan_id, path, file_name):
...
+    and apply the scan.
+    """
+    logger = logging.getLogger(__name__)
+    logger.info('git repo detected')
+
+    clean_url = ''.join(e for e in url if e.isalnum())
+
+    dir_name = clean_url
+
+    home_path = expanduser("~")
```
Comment (pombredanne) [inline scanapp/tasks.py]: Saving under a user dir down not make sense. We need a well defined configured place.

File: scanapp/tasks.py [hunk 5]
```diff
@@ -57,6 +62,42 @@ def scan_code_async(url, scan_id, path, file_name):
...
+    os.mkdir(dir_name)
+
+    repo = git.Repo.init(dir_name)
+    origin = repo.create_remote('origin', url)
+    origin.fetch()
+    origin.pull(origin.refs[0].remote_head)
+
+    logger.info('Done ! Remote repository cloned')
+
+    filename = home_path + '/' + clean_url + '/'
```
Comment (pombredanne) [inline scanapp/tasks.py]: use os.path.join.

File: scanapp/tasks.py [hunk 6]
```diff
@@ -82,11 +123,11 @@ def save_results_to_db(scan_id, json_data):
         scancode_version=json_data['scancode_version'],
     )
 
-# logic to calculate total_error
-#    total_errors = 0
-#    for a_file in json_data['files']:
-#        for error in a_file['scan_errors']:
-#            total_errors = total_errors + 1
+    # logic to calculate total_error
```
Comment (pombredanne) [inline scanapp/tasks.py]: Do not keep commented out code

File: scanapp/tasks.py [hunk 7]
```diff
@@ -82,11 +123,11 @@ def save_results_to_db(scan_id, json_data):
...
+    #    total_errors = 0
+    #    for a_file in json_data['files']:
+    #        for error in a_file['scan_errors']:
+    #            total_errors = total_errors + 1
 
     for a_file in json_data['files']:
```
Comment (pombredanne) [inline scanapp/tasks.py]: a_file is not a proper variable name. Call it file or scanned_file

Comment (pombredanne) [reply scanapp/tasks.py]: do not prefix with a_ elsewhere too.

File: scanapp/tasks.py [hunk 8]
```diff
@@ -188,3 +229,17 @@ def fill_unfilled_scan_model(scan, files_count, scancode_notice, scancode_versio
...
+
+
+def parse_url(URL):
+    """
+        Parses the URL and checks if it's a git URL. If it is a git URL then the flag is set to 1.
+    """
+    flag = 0
+    g = GitURL(URL)
+
+    try:
```
Comment (pombredanne) [inline scanapp/tasks.py]: Why this try block?

File: scanapp/views.py [hunk 9]
```diff
@@ -22,32 +22,32 @@
 #  Visit https://github.com/nexB/scancode-server/ for support and download.
 
 import json
+import logging
 import os
 import subprocess
```
Comment (pombredanne) [inline scanapp/views.py:27]: sort imports

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks!
See my comments.... Are you properly synced and rebased on the latest develop?

File: requirements.txt (hunk 1 above)
Comment (RajuKoushik) [reply requirements.txt]: We need one. I'll update.

File: scanapp/tasks.py (hunk 2 above)
Comment (RajuKoushik) [reply scanapp/tasks.py:26]: Done!

File: scanapp/tasks.py (hunk 4 above)
Comment (RajuKoushik) [reply scanapp/tasks.py]: Sure !

File: scanapp/tasks.py (hunk 8 above)
Comment (RajuKoushik) [reply scanapp/tasks.py]: The function which I am using to check whether it is a git url or not would return null in the case of a random absurd url but we to return the 'flag' variable every time. So the try block.

File: scanapp/views.py (hunk 9 above)
Comment (RajuKoushik) [reply scanapp/views.py:27]: Sure ! Done.

File: scanapp/tasks.py (hunk 6 above)
Comment (RajuKoushik) [reply scanapp/tasks.py]: 👍

File: scanapp/tasks.py (hunk 7 above)
Comment (RajuKoushik) [reply scanapp/tasks.py]: Done !

File: scanapp/tasks.py [hunk 10]
```diff
@@ -24,21 +24,27 @@
 from __future__ import absolute_import, unicode_literals
 
 import json
+import logging
+import os
+import shutil
 import subprocess
-import requests
 from datetime import datetime
```
Comment (pombredanne) [inline scanapp/tasks.py]: This is not sorted?  datetime comes before json, right?

File: scanapp/tasks.py [hunk 11]
```diff
@@ -57,6 +63,44 @@ def scan_code_async(url, scan_id, path, file_name):
...
+def scan_code_async_final(url, scan_id, path):
+    """
+    Create and initialise the git repository at a certain path and clone the git repo using the 'url'
+    and then get the scan done asynchronously.
+    """
+    logger = logging.getLogger(__name__)
+    logger.info('git repo detected')
+
+    home_path = expanduser("~")
+    os.chdir(home_path)
```
Comment (pombredanne) [inline scanapp/tasks.py]: Why do you use these chdir?

File: scanapp/tasks.py [hunk 12]
```diff
@@ -57,6 +63,44 @@ def scan_code_async(url, scan_id, path, file_name):
...
+
+    os.chdir(path)
+
+    print os.curdir
```
Comment (pombredanne) [inline scanapp/tasks.py]: why leaving print statements?

File: scanapp/tasks.py [hunk 13]
```diff
@@ -57,6 +63,44 @@ def scan_code_async(url, scan_id, path, file_name):
...
+
+    clean_url = ''.join(e for e in url if e.isalnum())
```
Comment (pombredanne) [inline scanapp/tasks.py]: Why do you do this?

File: scanapp/tasks.py [hunk 14]
```diff
@@ -57,6 +63,44 @@ def scan_code_async(url, scan_id, path, file_name):
...
+    dir_name = clean_url
+
+    if os.path.isdir(dir_name):
+        shutil.rmtree(dir_name)
```
Comment (pombredanne) [inline scanapp/tasks.py]: Never ever delete things... Why are you doing this?

File: scanapp/tasks.py [hunk 15]
```diff
@@ -57,6 +63,44 @@ def scan_code_async(url, scan_id, path, file_name):
...
+
+    os.mkdir(dir_name)
```
Comment (pombredanne) [inline scanapp/tasks.py]: The whole way you handle directories is broken.
You are doing a checkout. You want to get a temp directory and nothing else. Use the scancode.commoncode.fileutils code at https://github.com/nexB/scancode-toolkit/blob/22bf88df19768dd207973ed2db91249822a0c446/src/commoncode/fileutils.py#L120 and use a well defined base dir. Each new clone should go in a new temp directory. Never deleted

File: scanapp/tasks.py [hunk 16]
```diff
@@ -57,6 +63,44 @@ def scan_code_async(url, scan_id, path, file_name):
...
+
+    repo = git.Repo.init(dir_name)
+    origin = repo.create_remote('origin', url)
+    origin.fetch()
+    origin.pull(origin.refs[0].remote_head)
```
Comment (pombredanne) [inline scanapp/tasks.py]: What is the magic refs[0]?

File: scanapp/tasks.py [hunk 17]
```diff
@@ -57,6 +63,44 @@ def scan_code_async(url, scan_id, path, file_name):
...
+
+    logger.info('Done ! Remote repository cloned')
+
+    filename = home_path + '/' + path + dir_name + '/'
```
Comment (pombredanne) [inline scanapp/tasks.py]: Use os.path.join. Not this

File: scanapp/tasks.py [hunk 18]
```diff
@@ -188,3 +226,17 @@ def fill_unfilled_scan_model(scan, files_count, scancode_notice, scancode_versio
     scan.scancode_version = scancode_version
     scan.save()
     return scan
+
+
+def parse_url(URL):
+    """
+        Parses the URL and checks if it's a git URL. If it is a git URL then the flag is set to 1.
```
Comment (pombredanne) [inline scanapp/tasks.py]: Incorrect indentation

Comment (pombredanne) [reply scanapp/tasks.py]: What is flag? is this something that is returned?

File: scanapp/tasks.py [hunk 19]
```diff
@@ -188,3 +226,17 @@ def fill_unfilled_scan_model(scan, files_count, scancode_notice, scancode_versio
     scan.scancode_version = scancode_version
     scan.save()
     return scan
+
+
+def parse_url(URL):
```
Comment (pombredanne) [inline scanapp/tasks.py]: do not use uppercase for variable and arg names

File: scanapp/tasks.py (hunk 8 above)
Comment (pombredanne) [reply scanapp/tasks.py]: use a boolean instead. None evals to False alright.  You could replace this whole function with
`return GitURL(url).is_a('github')`
And if this is a onliner, why a function? also the doc string is misleading if you only handle github urls and not generally git-related urls

File: scanapp/views.py [hunk 20]
```diff
@@ -161,3 +123,79 @@ def post(self, request):
                 }
             )
         )
+
```
Comment (pombredanne) [inline scanapp/views.py]: Why did you move this code around? I cannot review which changes you made there

File: scanapp/views.py [hunk 21]
```diff
@@ -161,3 +123,79 @@ def post(self, request):
...
+                # logic to check how many files are already present for the scan
+                dir_list = list()
+                dir_list = os.listdir(path)
+                file_name = ''
+
+                if len(dir_list) == 0:
+                    file_name = '1'
+                else:
+                    dir_list.sort()
+                    file_name = str(1 + int(dir_list[-1]))
```
Comment (pombredanne) [inline scanapp/views.py]: What is this code? I do not get what you are doing or trying to do with this weird thing. What is it?

File: scanapp/views.py [hunk 22]
```diff
@@ -161,3 +123,79 @@ def post(self, request):
...
+
+                scan_directory = file_name
+                scan_id = create_scan_id(user, url, scan_directory, scan_start_time)
+                scan_code_async.delay(url, scan_id, path, file_name)
+                return HttpResponseRedirect('/resultscan/' + str(scan_id))
```
Comment (pombredanne) [inline scanapp/views.py]: what is str(scan_id) ? why do not you use a proper attribute?

File: scanapp/views.py [hunk 23]
```diff
@@ -161,3 +123,79 @@ def post(self, request):
...
+                    user = None
+                else:
+                    path = 'media/user/' + str(request.user) + '/url/'
+                    user = request.user
+
+                scan_start_time = datetime.now()
+                subprocess.call(['mkdir', '-p', path])
+
+                # logic to check how many files are already present for the scan
+                dir_list = list()
```
Comment (pombredanne) [inline scanapp/views.py]: This whole code block is weird and arcane. What are you trying to do ?

File: scanapp/views.py [hunk 24]
```diff
@@ -161,3 +123,79 @@ def post(self, request):
...
+
+                # return the response as HttpResponse
+                return HttpResponseRedirect('/resultscan/' + str(scan_id))
+
+            else:
+
+                url = request.POST['url']
+
+                # different paths for both anonymous and registered users
+                if (str(request.user) == 'AnonymousUser'):
```
Comment (pombredanne) [inline scanapp/views.py]: why the str()?

File: scanapp/views.py [hunk 25]
```diff
@@ -161,3 +123,79 @@ def post(self, request):
...
+                    path = 'media/AnonymousUser/url/'
+                    user = None
+                else:
+                    path = 'media/user/' + str(request.user) + '/url/'
```
Comment (pombredanne) [inline scanapp/views.py]: why str()?

Comment (pombredanne) [reply scanapp/views.py]: Also do not join path-lile arguments with +, use posixpath.join or url-related functions instead or even a simple '/'.join() would be better

File: scanapp/views.py [hunk 26]
```diff
@@ -161,3 +123,79 @@ def post(self, request):
...
+
+                if (str(request.user) == 'AnonymousUser'):
+                    path = 'media/AnonymousUser/url/'
+                    user = None
+                else:
+                    path = 'media/user/' + str(request.user) + '/url/'
+                    user = request.user
+
+                scan_start_time = datetime.now()
+                subprocess.call(['mkdir', '-p', path])
```
Comment (pombredanne) [inline scanapp/views.py]: Are you serious? Calling a subprocess to create a directory? did you check the Python doc?

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks.
See my comments... There are quire a few problems that need some serious rework.

File: scanapp/tasks.py (hunk 10 above)
Comment (RajuKoushik) [reply scanapp/tasks.py]: These were actually automatically sorted by my PyCharm editor according to pep8 guidelines.

File: scanapp/tasks.py (hunk 11 above)
Comment (RajuKoushik) [reply scanapp/tasks.py]: I am actually moving to the child directory and creating a new temp directory over there.

File: scanapp/tasks.py (hunk 13 above)
Comment (RajuKoushik) [reply scanapp/tasks.py]: I am trying to give the directory a name using it's url. So cleaning the scanned URL and using it as a name.

File: scanapp/tasks.py (hunk 14 above)
Comment (RajuKoushik) [reply scanapp/tasks.py]: Sure ! I'll try a better way out if the directory already exists.

File: scanapp/models.py [hunk 27]
```diff
@@ -59,7 +59,9 @@ def __str__(self):
     )
     scan_directory = models.CharField(
         max_length=200,
```
Comment (pombredanne) [inline scanapp/models.py]: Are you sure a path cannot be longer than this?

File: scanapp/tasks.py [hunk 28]
```diff
@@ -24,9 +24,11 @@
 from __future__ import absolute_import, unicode_literals
 
 import json
+import logging
 import requests
 import subprocess
 
+from giturl import *
```
Comment (pombredanne) [inline scanapp/tasks.py]: never import * and sort your imports: g comes after d

File: scanapp/tasks.py [hunk 29]
```diff
@@ -57,6 +59,20 @@ def scan_code_async(url, scan_id, path, file_name):
 
 
 @app.task
+def handle_special_urls(url, scan_id, path, host):
+    """
+    Create and initialise the git repository at a certain path and clone the git repo using `url`
+    and then get the scan done.
+    """
+    if host == 'github.com':
+        logger = logging.getLogger(__name__)
```
Comment (pombredanne) [inline scanapp/tasks.py]: The logger creation should be top level. Not inside a function

File: scanapp/tasks.py [hunk 30]
```diff
@@ -57,6 +59,20 @@ def scan_code_async(url, scan_id, path, file_name):
...
+        logger.info('git repo detected')
```
Comment (pombredanne) [inline scanapp/tasks.py]: a  log entry should contain more details, e.g. information about the args used (here the host at least)

File: scanapp/views.py [hunk 31]
```diff
@@ -31,21 +32,24 @@
 from django.http import HttpResponse
 from django.http import HttpResponseRedirect
 from django.shortcuts import render
+
 from django.utils import timezone
 from django.views import View
 from django.views.decorators.csrf import csrf_exempt
 from django.views.generic.base import TemplateView
 from django.views.generic.edit import FormView
+from giturl import *
```
Comment (pombredanne) [inline scanapp/views.py]: No import * please

File: scanapp/views.py [hunk 32]
```diff
@@ -31,21 +32,24 @@
 from django.http import HttpResponse
 from django.http import HttpResponseRedirect
 from django.shortcuts import render
+
```
Comment (pombredanne) [inline scanapp/views.py]: Empty line not needed here?

File: scanapp/views.py [hunk 33]
```diff
@@ -31,21 +32,24 @@
...
 from django.utils import timezone
 from django.views import View
 from django.views.decorators.csrf import csrf_exempt
 from django.views.generic.base import TemplateView
 from django.views.generic.edit import FormView
+from giturl import *
+from rest_framework.authtoken.models import Token
```
Comment (pombredanne) [inline scanapp/views.py]: No empty line here

File: scanapp/views.py [hunk 34]
```diff
@@ -31,21 +32,24 @@
...
 from scanapp.forms import UrlScanForm
-
 from scanapp.models import Scan
 
 from scanapp.tasks import apply_scan_async
 from scanapp.tasks import create_scan_id
+from scanapp.tasks import handle_special_urls
 from scanapp.tasks import scan_code_async
 
 from django.views.decorators.csrf import csrf_exempt
```
Comment (pombredanne) [inline scanapp/views.py]: Django imports should go with the thirdparty, before you scanapp import (e.g. 4 blocks: 1. `__*` imports, 2. stdlib imports, 3. thirdparty imports, 4. own code imports)

File: scanapp/views.py [hunk 35]
```diff
@@ -178,6 +145,50 @@ def post(self, request):
...
+    form_class = UrlScanForm
+
+    def post(self, request, *args, **kwargs):
+        form = self.form_class(request.POST)
+
+        if form.is_valid():
+            url = request.POST['url']
+            logger = logging.getLogger(__name__)
+
+            if str(request.user) == 'AnonymousUser':
```
Comment (pombredanne) [inline scanapp/views.py]: Do not use str() ... not here and almost nowhere. Use the attribute (e.g. here likely reques.user.name ?)

File: scanapp/views.py [hunk 36]
```diff
@@ -178,6 +145,50 @@ def post(self, request):
...
+                user = request.user
+
+            scan_start_time = timezone.now()
+            git_url_parser = GitURL(url)
+
+            if git_url_parser.host == 'github.com':
+                file_name = git_url_parser.repo
+                scan_directory = file_name
+                scan_id = create_scan_id(user, url, scan_directory, scan_start_time)
+                path = '/'.join([''.join([path, str(scan_id)]), file_name])
```
Comment (pombredanne) [inline scanapp/views.py]: no str()

**Review by pombredanne — CHANGES_REQUESTED** [review]
See my comments... This would need some tests too :P

File: scanapp/tasks.py (hunk 28 above)
Comment (singh1114) [reply scanapp/tasks.py]: It was given in the documentation of giturl. 
https://github.com/willyg302/giturl#usage

File: scanapp/models.py (hunk 27 above)
Comment (singh1114) [reply scanapp/models.py]: Ok, can we change it to 2000.

File: scanapp/tasks.py (hunk 29 above)
Comment (singh1114) [reply scanapp/tasks.py]: Yes! Done.

File: scanapp/views.py (hunk 35 above)
Comment (singh1114) [reply scanapp/views.py]: `"Scan.user" must be a "User" instance.` Error produced after removing `str`

File: scanapp/views.py (hunk 36 above)
Comment (singh1114) [reply scanapp/views.py]: `sequence item 1: expected string or Unicode, int found`
found this error after removing `str`. Is there any method to take care of this.

Comment (RajuKoushik) [reply scanapp/views.py]: Done ! Used the format() function instead.

File: scanapp/tasks.py [hunk 37]
```diff
@@ -24,10 +24,12 @@
 from __future__ import absolute_import, unicode_literals
 
 import json
-import requests
+import logging
 import subprocess
 
+import requests
```
Comment (pombredanne) [inline scanapp/tasks.py:30]: request starts with r and goes after g ;)

File: scanapp/views.py [hunk 38]
```diff
@@ -33,34 +34,23 @@
 from django.shortcuts import render
 from django.utils import timezone
 from django.views import View
-from django.views.decorators.csrf import csrf_exempt
 from django.views.generic.base import TemplateView
 from django.views.generic.edit import FormView
-
+from giturl import *
```
Comment (pombredanne) [inline scanapp/views.py]: No import * please. Use exactly what you need and no more

**Pull request merged**